chain/client/src/client_actor.rs

```diff
@@ -1117,42 +1117,58 @@ impl ClientActorInner {
-                )?;
-                let num_chunks = self
-                    .client
-                    .chunk_inclusion_tracker
-                    .num_chunk_headers_ready_for_inclusion(&epoch_id, &head.last_block_hash);
-                let shard_ids = self.client.epoch_manager.shard_ids(&epoch_id).unwrap();
-                let have_all_chunks = head.height == 0 || num_chunks == shard_ids.len();
+            if me != next_block_producer_account {
+                continue;
+            }
 
-                if self.client.doomslug.ready_to_produce_block(
-                    height,
-                    have_all_chunks,
-                    log_block_production_info,
-                ) {
-                    self.client
-                        .chunk_inclusion_tracker
-                        .record_endorsement_metrics(&head.last_block_hash, &shard_ids);
-                    if let Err(err) = self.produce_block(height, signer) {
-                        // If there is an error, report it and let it retry on the next loop step.
-                        error!(target: "client", height, "Block production failed: {}", err);
-                    } else {
-                        self.post_block_production();
-                    }
+            self.client.chunk_inclusion_tracker.prepare_chunk_headers_ready_for_inclusion(
+                prev_block_hash,
+                &mut self.client.chunk_endorsement_tracker,
+            )?;
+            let num_chunks = self
+                .client
+                .chunk_inclusion_tracker
+                .num_chunk_headers_ready_for_inclusion(&epoch_id, prev_block_hash);
+            let shard_ids = self.client.epoch_manager.shard_ids(&epoch_id).unwrap();
+            let have_all_chunks = head.height == 0 || num_chunks == shard_ids.len();
+
+            if self.client.doomslug.ready_to_produce_block(
+                height,
+                have_all_chunks,
+                log_block_production_info,
+            ) {
+                self.client
+                    .chunk_inclusion_tracker
+                    .record_endorsement_metrics(prev_block_hash, &shard_ids);
+                if let Err(err) = self.produce_block(height, signer) {
+                    // If there is an error, report it and let it retry on the next loop step.
+                    error!(target: "client", height, "Block production failed: {}", err);
+                } else {
+                    self.post_block_production();
                 }
             }
         }
 
+        let protocol_version = self.client.epoch_manager.get_epoch_protocol_version(&epoch_id)?;
+        if !ProtocolFeature::ProduceOptimisticBlock.enabled(protocol_version) {
+            return Ok(());
+        }
+        let optimistic_block_height = self.client.doomslug.get_timer_height();
+        if self.client.epoch_manager.get_block_producer(&epoch_id, optimistic_block_height)? != me {
```

Comment:
```suggestion
        if me != self.client.epoch_manager.get_block_producer(&epoch_id, optimistic_block_height)? {
```

super nit which IMO improves readability